fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/VectorIndexMaintainer.java

```diff
@@ -0,0 +1,493 @@
+
+        final ExecuteProperties executeProperties = scanProperties.getExecuteProperties();
+        final ScanProperties innerScanProperties = scanProperties.with(ExecuteProperties::clearSkipAndLimit);
+        final Subspace indexSubspace = getIndexSubspace();
+        final FDBStoreTimer timer = Objects.requireNonNull(state.context.getTimer());
+
+        //
+        // Skip-scan through the prefixes in a way that we only consider each distinct prefix. That skip scan
+        // forms the outer of a join with an inner that searches the R-tree for that prefix using the
+        // spatial predicates of the scan bounds.
+        //
+        return RecordCursor.flatMapPipelined(prefixSkipScan(prefixSize, timer, vectorIndexScanBounds, innerScanProperties),
+                        (prefixTuple, innerContinuation) -> {
+                            final Subspace hnswSubspace;
+                            if (prefixTuple != null) {
+                                Verify.verify(prefixTuple.size() == prefixSize);
+                                hnswSubspace = indexSubspace.subspace(prefixTuple);
+                            } else {
+                                hnswSubspace = indexSubspace;
+                            }
+
+                            if (innerContinuation != null) {
+                                final RecordCursorProto.VectorIndexScanContinuation parsedContinuation =
+                                        Continuation.fromBytes(innerContinuation);
+                                final ImmutableList.Builder<IndexEntry> indexEntriesBuilder = ImmutableList.builder();
+                                for (int i = 0; i < parsedContinuation.getIndexEntriesCount(); i ++) {
+                                    final RecordCursorProto.VectorIndexScanContinuation.IndexEntry indexEntryProto =
+                                            parsedContinuation.getIndexEntries(i);
+                                    indexEntriesBuilder.add(new IndexEntry(state.index,
+                                            Tuple.fromBytes(indexEntryProto.getKey().toByteArray()),
+                                            Tuple.fromBytes(indexEntryProto.getValue().toByteArray())));
+                                }
+                                final ImmutableList<IndexEntry> indexEntries = indexEntriesBuilder.build();
+                                return new ListCursor<>(indexEntries, parsedContinuation.getInnerContinuation().toByteArray())
+                                        .mapResult(result ->
+                                                result.withContinuation(new Continuation(indexEntries, result.getContinuation())));
+                            }
+
+                            final HNSW hnsw = new HNSW(hnswSubspace, getExecutor(), getConfig(),
+                                    OnWriteListener.NOOP, new OnRead(timer));
+                            final ReadTransaction transaction = state.context.readTransaction(true);
```

Comment:
Why is this reading at snapshot isolation level?

Reply:
I asked @g31pranjal about this in the MD-code path.